### PR TITLE
Task #20918 (Improve PNG filesize on save)

### PR DESCRIPTION
**Pull request by Bujacikk**
Hi, I was working on task: https://github.com/OpenRA/OpenRA/issues/20918

I little changed Save method for 8bbp, 24bbp and alpha channel. I hope it is good. I added DeflaterOutputStream with max compression.
I add some unit tests too. I know it was not in the task, but I am doing it as a school task and we had to add some tests, so I hope It is not a big problem. If it is, I can remove them. 

PS: I am sorry for git messages. I read conditions too late.

### CODE REVIEW

**Comment by RoosterDragon** [comment]
Extra unit tests are great, no issues with that.

For the commits, you might find the following useful to condense your changes into a single commit, and for updating the commit message. We would usually ask that the commit message is something that describes the change.
- https://stackoverflow.com/questions/5189560/how-do-i-squash-my-last-n-commits-together
- https://stackoverflow.com/questions/179123/how-to-modify-existing-unpushed-commit-messages

As both of those questions note, you will need to force push those changes, so that your PR is updated.

Thanks for your contribution.

**Comment by Bujacikk** [comment]
> Extra unit tests are great, no issues with that.
> 
> For the commits, you might find the following useful to condense your changes into a single commit, and for updating the commit message. We would usually ask that the commit message is something that describes the change.
> 
> * https://stackoverflow.com/questions/5189560/how-do-i-squash-my-last-n-commits-together
> * https://stackoverflow.com/questions/179123/how-to-modify-existing-unpushed-commit-messages
> 
> As both of those questions note, you will need to force push those changes, so that your PR is updated.
> 
> Thanks for your contribution.

Done

File: OpenRA.Game/FileFormats/Png.cs [hunk 1]
```diff
@@ -350,13 +351,13 @@ public byte[] Save()
...
 				{
-					using (var compressed = new DeflaterOutputStream(data))
+					using (var compressed = new DeflaterOutputStream(data, new Deflater(Deflater.BEST_COMPRESSION)))
 					{
 						var rowStride = Width * PixelStride;
 						for (var y = 0; y < Height; y++)
 						{
-							// Write uncompressed scanlines for simplicity
-							compressed.WriteByte(0);
+							var filterType = (byte)0; // Assuming no filtering for simplicity
```
Comment (abcdefg30) [inline OpenRA.Game/FileFormats/Png.cs]: Our convention is to put comments on their own line:
```suggestion
							// Assuming no filtering for simplicity
							var filterType = (byte)0;
```

File: OpenRA.Test/OpenRA.Game/PngTest.cs [hunk 2]
```diff
@@ -0,0 +1,145 @@
+﻿using System.IO;
```
Comment (abcdefg30) [inline OpenRA.Test/OpenRA.Game/PngTest.cs]: Can you please add the header here?

```
#region Copyright & License Information
/*
 * Copyright (c) The OpenRA Developers and Contributors
 * This file is part of OpenRA, which is free software. It is made
 * available to you under the terms of the GNU General Public License
 * as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version. For more
 * information, see COPYING.
 */
#endregion
```

File: OpenRA.Test/OpenRA.Game/PngTest.cs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+	[Test]
+	public void PngConstructor_HeaderNotFirst_ThrowsInvalidDataException()
+	{
+		// Arrange
+		var invalidPngData = new byte[]
+		{
+		0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a, 
+		0x00, 0x00, 0x00, 0x00, 0x49, 0x45, 0x4e, 0x44, 
+		0x00, 0x00, 0x00, 0x00,
```
Comment (abcdefg30) [inline OpenRA.Test/OpenRA.Game/PngTest.cs]:
```suggestion
			0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a, 
			0x00, 0x00, 0x00, 0x00, 0x49, 0x45, 0x4e, 0x44, 
			0x00, 0x00, 0x00, 0x00, 
```

here and below.

File: OpenRA.Test/OpenRA.Game/PngTest.cs [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+		// Arrange
+		var colors = Enumerable.Range(0, 256).Select(i => Color.FromArgb(i, i, i)).ToArray();
+		var png = new Png(new byte[10 * 20], SpriteFrameType.Indexed8, 10, 20, colors);
+
+		// Act
+		var result = png.Save();
+
+		// Assert
+		var colorTypeByte = result[25]; // Byte at index 25 contains color type information
+		Assert.That(colorTypeByte, Is.EqualTo(0x03)); // 0x03 represents Indexed8 with a palette
```
Comment (abcdefg30) [inline OpenRA.Test/OpenRA.Game/PngTest.cs]: Hm, maybe something like

```suggestion
		// Assert:
		// The byte at index 25 contains color type information
		// and 0x03 represents Indexed8 with a palette
		var colorTypeByte = result[25];
		Assert.That(colorTypeByte, Is.EqualTo(0x03));
```

here and below?

**Review by abcdefg30 — COMMENTED** [review]
Code changes seem simple enough. 👍 Some minor/style nits.

**Comment by PunkPun** [comment]
Ping, should the PR be taken over?

**Review by PunkPun — APPROVED** [review]
Force pushed fixes to style violations. LGTM

**Comment by PunkPun** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d2e01d157a224ab6b4f92a038868738e97feab85...92bb424315d9c484e7a89e031efeca1b3a5968b2)

**Pull request merged**